src/app/core/services/label.service.ts

```diff
@@ -110,16 +112,27 @@ export class LabelService {
   }
 
   /**
-   * Converts the (color) hex value into RGB format
-   * @param hex: the hex value
+   * Choose black or white text color depending on background color
+   * @param bgColor: the color code of the background
    */
-  hexToRgb(hex: string) {
-    const rgbResult = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
-    return rgbResult ? {
-      r: parseInt(rgbResult[1], 16),
-      g: parseInt(rgbResult[2], 16),
-      b: parseInt(rgbResult[3], 16)
-    } : null;
+  pickTextColorBasedOnBgColor(bgColor: string) {
+    const lightColor = 'FFFFFF'; // Light text color (white)
+    const darkColor = '000000'; // Dark text color (black)
+    const color = (bgColor.charAt(0) === '#') ? bgColor.substring(1, 7) : bgColor;
+    const r = parseInt(color.substring(0, 2), 16);
+    const g = parseInt(color.substring(2, 4), 16);
+    const b = parseInt(color.substring(4, 6), 16);
+    const uiColors = [r / 255, g / 255, b / 255];
+    const c = uiColors.map((col) => {
+      if (col <= 0.03928) {
+        return col / 12.92;
+      }
+      return Math.pow((col + 0.055) / 1.055, 2.4);
+    });
+    // Calculate the luminance of the background color
+    const L = (0.2126 * c[0]) + (0.7152 * c[1]) + (0.0722 * c[2]);
+    // Higher threshold will result in more labels with light color text
+    return (L > BGCOLOR_THRESHOLD) ? darkColor : lightColor;
   }
```

Comment:
Perhaps extract a utility function `isDarkColor` that returns true if the given color is dark?